### ISSUE TITLE
`poetry lock --check` doesn't warn of missmatch in versions of local dependencies and their transitive dependencies

### ISSUE THREAD

**Issue by ena-cradle** [python-poetry/poetry#8205]
- **Poetry version**: 1.5.1 (curl), also tried [installing @main with pipx](https://python-poetry.org/docs/#installing-with-pipx)
- **Python version**: 3.10
- **OS version and name**: macOS 12.5
- **pyproject.toml**: [gist](https://gist.github.com/ena-cradle/5b7d4a73b83aa738cdf071bb39389cf6)

- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below.

## Issue
Consider the following project structure:
```
|- A
|  |- a
|  |  |- __init__.py
|  |  |- ...
|  |
|  |- pyproject.toml
|
|- B
|  |- b
|  |  |- __init__.py
|  |  |- ...
|  |
|  |- pyproject.toml
```

and `B` depends on `A` (see [gist](https://gist.github.com/ena-cradle/5b7d4a73b83aa738cdf071bb39389cf6) for `pyproject.toml` content).
  
`B`'s `poetry.lock` also contains packages `A` depends on (so not directly stated in `B`'s `pyproject.toml`), **but** if a `A` changes a version of its dependency, running `poetry lock --check --directory B` will not register that the lockfile is not consistent anymore. However, running `poetry lock --no-update --directory B` will change the lockfile. 

The same happens in case `A` bumps its version 

### Expected output
In this case I'd expect `poetry lock --check --directory B` to report the inconsistency.

## Full example for transitive dependencies
Full example starting from the `pyproject.toml` from the gist:
```bash
>> poetry lock --directory A
Updating dependencies
Resolving dependencies... (0.3s)

Writing lock file

>> poetry lock --directory B
Updating dependencies
Resolving dependencies... (0.3s)

Writing lock file 

>> poetry add "plotly@<5.15" --directory A
Updating dependencies
Resolving dependencies... (0.1s)

Package operations: 3 installs, 0 updates, 0 removals

 • Installing packaging (23.1)
 • Installing tenacity (8.2.2)
 • Installing plotly (5.14.1)

Writing lock file

>> poetry lock --check --directory A      
poetry.lock is consistent with pyproject.toml.

>> poetry lock --check --directory B      
poetry.lock is consistent with pyproject.toml.

>> poetry lock --directory B        
Updating dependencies
Resolving dependencies... (0.2s)

Writing lock file
```

diff in `B`'s lockfile: [gist](https://gist.github.com/ena-cradle/d9f63f1084e2eedd66f756b7f889cf41) 
TL;DR the entry for package `a` changed and the entry for `plotly` changed to downgrade the version.

## Full example for local dependency version bump
Full example starting from the `pyproject.toml` from the gist:
```bash
>> poetry lock --directory A
Updating dependencies
Resolving dependencies... (0.3s)

Writing lock file

>> poetry lock --directory B
Updating dependencies
Resolving dependencies... (0.3s)

Writing lock file 

>>  poetry version patch --directory A
Bumping version from 0.1.0 to 0.1.1

>> poetry lock --check --directory A      
poetry.lock is consistent with pyproject.toml.

>> poetry lock --check --directory B      
poetry.lock is consistent with pyproject.toml.

>> poetry lock --directory B        
Updating dependencies
Resolving dependencies... (0.2s)

Writing lock file
```

diff in `B`'s lockfile: 

before:
```toml
[[package]]
name = "a"
version = "0.1.0"
description = ""
optional = false
python-versions = "^3.10"
files = []
develop = true
```

after:
```toml
[[package]]
name = "a"
version = "0.1.1"
description = ""
optional = false
python-versions = "^3.10"
files = []
develop = true
```
